### PR TITLE
Assignment6 - Joshua Willis

### PR DESCRIPTION
**Pull request by WobblyRunner**
I took a little time to go above-and-beyond with my implementation here. This solution allows for complex expressions to be evaluated while still abiding by the TryParse pattern. I updated the README.md as a reference to see what I included and did include. I did the refactoring extra credit but still included my base implementation and tests for Calculator.cs as well.

### CODE REVIEW

**Comment by Afterlyte** [comment]
I will review this

File: Calculate/Calculate.Tests/GenericCalculatorTests.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+		GenericCalculator<float> calculator = new GenericCalculator<float>();
+		const string expression = "1 + 3 * 4";
+		const float expected = (1F + 3F) * 4F;
+		bool success;
+
+		success = calculator.TryCalculate(expression, out float actual);
+
+		Assert.IsTrue(success);
+		Assert.AreEqual<float>(expected, actual);
+	}
```
Comment (Afterlyte) [inline Calculate/Calculate.Tests/GenericCalculatorTests.cs:23]: I really appreciate the formatting of your tests, very readable and consistent

File: Calculate/Calculate/Calculator.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+			{ '-', Subtract },
+			{ '*', Multiply },
+			{ '/', Divide },
+		};
+		Operators = MathematicalOperations.Keys.ToArray();
+	}
+
+	public bool TryCalculate(string expression, out int result)
+	{
+		int index = expression.LastIndexOfAny(Operators);
```
Comment (Afterlyte) [inline Calculate/Calculate/Calculator.cs:29]: I like this way of dynamically determining the last used operator. This looks a lot more reliable for more dynamic solutions like this

File: Calculate/Calculate/Calculator.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+		if (index > -1)
+		{
+			if (int.TryParse(expression[(index+1)..], out int right)
+				&& TryCalculate(expression[..index], out int left))
```
Comment (Afterlyte) [inline Calculate/Calculate/Calculator.cs:33]: Pattern matching is never used which is one of the requirements

File: Calculate/Calculate.Tests/CalculatorTests.cs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+		Calculator.Add = (lhs, rhs) => lhs + rhs;
+		Calculator.Subtract = (lhs, rhs) => lhs - rhs;
+		Calculator.Multiply = (lhs, rhs) => lhs * rhs;
+		Calculator.Divide = (lhs, rhs) => lhs / rhs;
+		TestCalculator = new Calculator();
+	}
+
+	void TestCalculationGivenExpression(string expression, int expected)
+	{
+		bool success = TestCalculator.TryCalculate(expression, out int actual);
```
Comment (Afterlyte) [inline Calculate/Calculate.Tests/CalculatorTests.cs:25]: Program.ReadLine/WriteLine needs to be used in testing to test that the assigned methods are giving the expected behavior

File: Calculate/Directory.Build.props [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
```
Comment (Afterlyte) [inline Calculate/Directory.Build.props:3]: EnableNETAnalyzers missing from the working props file

**Review by Afterlyte — CHANGES_REQUESTED** [review]
## Instructions

- Create a *Console* project called "Calculate.". ✔
- Define a Program Class
  - Define two init-only setter properties, `WriteLine` and `ReadLine`, that contain delegates for writing a line of text and reading a line of text respectively ✔
  - Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
  - Set the default behavior for the `WriteLine` and `ReadLine` properties to invoke `System.Console` versions of the methods and add an empty default constructor. ✔
- Define a Calculator class ✔
  - Define static `Add`, `Subtract`, `Multiple`, and `Divide` methods that have two parameters and return a third parameter. ✔
  - Define a read-only property, `MathematicalOperations`, of type `System.Collections.Generics.IReadOnlyDictionary<TKey,TValue>` that:
    - is initialized to a `System.Collections.Generics.Dictionary<<TKey,TValue>` instance that. ✔
      - Uses `char` for the key corresponding to the operators +, -, *, and /. ✔
      - Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
  - Implement a `TryCalculate` method following "TryParse" pattern ✔
    - Valid `calculation` expressions include such strings as "3 + 4", "42 - 2", etc. ✔
     - **You went a little above that but hey the standard requirement still works**
    - If there is no whitespace around the operator, you can assume the `calculation` is invalid and return false. Similarly if the operands are not integers. ✔
      - **Technically you didn't do this but because you handled whitespace properly**
    - Use `string.Split()`, pattern matching, logical and operators to parse the string in their entirety ❌
      - **It doesn't look like you used pattern matching at all on the expressions, but instead did indexing on the string** 
    - Index into the `MathematicalOperations` method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
- Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
- Be sure to use the `WriteLine`/`ReadLine` properties on `Program` for testing the input and output of your program. ❌
  - **program.ReadLine/WriteLine was used in the actual main method, but never used during testing.**

## Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

- Refactor the redirect portion of the `Program` class into 'ProgramBase` ❌
- Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects ❌
- Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0) ✔

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file. ✔
  - **There's a duplicate file but the correct one still works**
- Place all shared project items into a `Directory.Build.targets` file. 
- nullable reference types is enabled  ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ❌
  - **NET analyzers are not enabled on the correct version of the file**
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional)   ✔
- and enabled .NET analyzers for both projects ❌
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ✔
- All of the above should be unit tested ❌
  - **Missing unit tests on using Program.ReadLine/WriteLine when doing the calculations but good other than that.**
- Choose simplicity over complexity ✔
  - **Well, debatable but you made complex functionality so it counts**

**Overall really good and clever extra functionality, mainly missing some unit tests and missing the pattern matching on the expression input**

**Comment by JessicaH46** [comment]
Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
You went a little above that but hey the standard requirement still works
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Technically you didn't do this but because you handled whitespace properly
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ❌
It doesn't look like you used pattern matching at all on the expressions, but instead did indexing on the string
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ❌✔?
program.ReadLine/WriteLine was used in the actual main method, but never used during testing.

I think you did really good on your code, most of the main components re done from what I saw. I think some of your fundamentals are missing but those are quick fixes! You also have a duplicte file for your Directory build props. Also for one of the requirements, readline writeline, I think you might need to add your Console.ReadLine here
`public Action<string> WriteLine { get; init; } = Console.WriteLine;
  public Func<string> ReadLine { get; init; } = Console.ReadLine;`
I had it like yours at first and it didn't work properly for mine at least. Everything else looks good, great work!

File: Calculate/Calculate/Calculator.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+		{
+			{ '+', Add },
+			{ '-', Subtract },
+			{ '*', Multiply },
+			{ '/', Divide },
+		};
+		Operators = MathematicalOperations.Keys.ToArray();
+	}
+
+	public bool TryCalculate(string expression, out int result)
```
Comment (vicdotexe) [inline Calculate/Calculate/Calculator.cs:27]: TryCalculate("FiveMinusFive", out int result);
Assert.AreEqual(0,result);
//what kind of magic is this?

**Review by mmwoodfo — COMMENTED** [review]
Awesome job, this looks great!

**Pull request closed without merging**